src/surface_header.F90

```diff
@@ -0,0 +1,946 @@
+    real(8) :: y0
+    real(8) :: z0
+    real(8) :: r
+  contains
+    procedure :: evaluate => x_cylinder_evaluate
+    procedure :: distance => x_cylinder_distance
+    procedure :: normal => x_cylinder_normal
+  end type SurfaceXCylinder
+
+  type, extends(Surface) :: SurfaceYCylinder
+    ! (x - x0)^2 + (z - z0)^2 = R^2
+    real(8) :: x0
+    real(8) :: z0
+    real(8) :: r
+  contains
+    procedure :: evaluate => y_cylinder_evaluate
+    procedure :: distance => y_cylinder_distance
+    procedure :: normal => y_cylinder_normal
+  end type SurfaceYCylinder
+
+  type, extends(Surface) :: SurfaceZCylinder
+    ! (x - x0)^2 + (y - y0)^2 = R^2
+    real(8) :: x0
+    real(8) :: y0
+    real(8) :: r
+  contains
+    procedure :: evaluate => z_cylinder_evaluate
+    procedure :: distance => z_cylinder_distance
+    procedure :: normal => z_cylinder_normal
+  end type SurfaceZCylinder
+
+  type, extends(Surface) :: SurfaceSphere
+    ! (x - x0)^2 + (y - y0)^2 + (z - z0)^2 = R^2
+    real(8) :: x0
+    real(8) :: y0
+    real(8) :: z0
+    real(8) :: r
+  contains
+    procedure :: evaluate => sphere_evaluate
+    procedure :: distance => sphere_distance
+    procedure :: normal => sphere_normal
+  end type SurfaceSphere
+
+  type, extends(Surface) :: SurfaceXCone
+    ! (y - y0)^2 + (z - z0)^2 = R^2*(x - x0)^2
+    real(8) :: x0
+    real(8) :: y0
+    real(8) :: z0
+    real(8) :: r2
+  contains
+    procedure :: evaluate => x_cone_evaluate
+    procedure :: distance => x_cone_distance
+    procedure :: normal => x_cone_normal
+  end type SurfaceXCone
+
+  type, extends(Surface) :: SurfaceYCone
+    ! (x - x0)^2 + (z - z0)^2 = R^2*(y - y0)^2
+    real(8) :: x0
+    real(8) :: y0
+    real(8) :: z0
+    real(8) :: r2
+  contains
+    procedure :: evaluate => y_cone_evaluate
+    procedure :: distance => y_cone_distance
+    procedure :: normal => y_cone_normal
+  end type SurfaceYCone
+
+  type, extends(Surface) :: SurfaceZCone
+    ! (x - x0)^2 + (y - y0)^2 = R^2*(z - z0)^2
+    real(8) :: x0
+    real(8) :: y0
+    real(8) :: z0
+    real(8) :: r2
+  contains
+    procedure :: evaluate => z_cone_evaluate
+    procedure :: distance => z_cone_distance
+    procedure :: normal => z_cone_normal
+  end type SurfaceZCone
+
+  abstract interface
+    pure function iEvaluate(this, xyz) result(f)
+      import Surface
+      class(Surface), intent(in) :: this
+      real(8), intent(in) :: xyz(3)
+      real(8) :: f
+    end function iEvaluate
+
+    pure function iDistance(this, xyz, uvw, coincident) result(d)
+      import Surface
+      class(Surface), intent(in) :: this
+      real(8), intent(in) :: xyz(3)
+      real(8), intent(in) :: uvw(3)
+      logical, intent(in) :: coincident
+      real(8) :: d
+    end function iDistance
+
+    pure function iNormal(this, xyz) result(uvw)
+      import Surface
+      class(Surface), intent(in) :: this
+      real(8), intent(in) :: xyz(3)
+      real(8) :: uvw(3)
+    end function iNormal
+  end interface
```

Comment:
I think of these guys more like interfaces in Java, where a class can implement an interface representing a collection of methods. So, syntactically `iDistance` seems to make more sense than `i_distance`, to me at least. All is well because the name of the actual bound procedure of Surface is still `distance`, and extension types are, e.g., `distance => x_plane_distance`.